codegen/core/src/main/java/software/amazon/smithy/java/codegen/generators/SchemaFieldOrder.java

```diff
@@ -0,0 +1,138 @@
+import software.amazon.smithy.model.loader.Prelude;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.ShapeType;
+import software.amazon.smithy.model.traits.UnitTypeTrait;
+import software.amazon.smithy.utils.CaseUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+@SmithyInternalApi
+public final class SchemaFieldOrder {
+
+    private static final EnumSet<ShapeType> EXCLUDED_TYPES = EnumSet.of(
+            ShapeType.SERVICE,
+            ShapeType.RESOURCE,
+            ShapeType.MEMBER,
+            ShapeType.OPERATION,
+            ShapeType.ENUM,
+            ShapeType.INT_ENUM);
+
+    private final List<List<SchemaField>> partitions;
+    private final Map<ShapeId, SchemaField> reverseMapping;
+    private final SymbolProvider symbolProvider;
+
+    public SchemaFieldOrder(Directive<?> directive, long partitionThreshold, SymbolProvider symbolProvider) {
+        var connectedShapes = new HashSet<>(directive.connectedShapes().values());
+        var index = TopologicalIndex.of(directive.model());
+        var allShapes = Stream.concat(index.getOrderedShapes().stream(), index.getRecursiveShapes().stream())
+                .filter(connectedShapes::contains)
+                .filter(s -> !s.hasTrait(SyntheticTrait.class))
+                .filter(s -> !EXCLUDED_TYPES.contains(s.getType()))
+                .filter(not(Prelude::isPreludeShape))
+                .collect(Collectors.toCollection(LinkedHashSet::new));
+        List<List<SchemaField>> computedPartitions = new ArrayList<>();
+        int curIndex = 0;
+        var curParition = new ArrayList<SchemaField>();
+        var curFieldNames = new HashSet<String>();
+        computedPartitions.add(curParition);
+        int curClassNumber = 0;
+        String curClassName = "Schemas";
+        for (var shape : allShapes) {
+            if (curIndex >= partitionThreshold) {
+                curIndex = 0;
+                curClassNumber++;
+                curClassName = "Schemas" + curClassNumber;
+                curFieldNames.clear();
+                curParition = new ArrayList<>();
+                computedPartitions.add(curParition);
+            }
+            var shapeFieldName = toSchemaName(shape);
+            if (curFieldNames.contains(shapeFieldName)) {
+                shapeFieldName = toFullQualifiedSchemaName(shape);
+            }
+            boolean isShapeRecursive = CodegenUtils.recursiveShape(directive.model(), shape);
+            var shapeField = new SchemaField(shape, shapeFieldName, curClassName, isShapeRecursive);
+            curParition.add(shapeField);
+            curFieldNames.add(shapeFieldName);
+            if (isShapeRecursive) {
+                curIndex++;
+            }
+            curIndex++;
+        }
+        this.partitions = computedPartitions.stream().filter(not(List::isEmpty)).toList();
```

Comment:
the big stream above makes some sense to use because it needs to concat two lists together and then slice it up. This one and the one below though could just be simpler constructs that don't allocate. Like this I think could be `computedPartitions.removeIf(List::isEmpty)`.